### PR TITLE
package up binaries in docker for pre-release testing

### PR DESCRIPTION
**Pull request by alexcb**
The Idea with this new target, is we could have it built on master, then we can use a script like this:

```
set -e

function get_latest_binary {
    docker pull alexcb132/earthlybinaries:latest
    docker create --name binaryhack alexcb132/earthlybinaries
    
    earthbinpath=/earth-linux-amd64
    if [ "$(uname)" == "Darwin" ]; then
    earthbinpath=/earth-darwin-amd64
    fi
    
    docker cp binaryhack:$earthbinpath ~/bin/earth-bin
    docker rm binaryhack
}

LAST=`cat /tmp/last-earth-check 2>/dev/null || echo 0`
NOW=`date +%s`
since=$(expr $NOW - $LAST)

if [ $since -ge 3600 ]; then
    echo pulling latest earth version
    get_latest_binary
    echo $NOW > /tmp/last-earth-check
fi

~/bin/earth-bin $@

```

to periodically check for latest pre-releases.

### CODE REVIEW

File: Earthfile [hunk 1]
```diff
@@ -146,6 +146,16 @@ earth-docker:
     COPY --build-arg VERSION=$TAG +earth/earth /usr/bin/earth
     SAVE IMAGE --push earthly/earth:$TAG
 
+# we abuse docker here to distribute our binaries
+prerelease-docker:
+    ARG EARTHLY_GIT_HASH
```
Comment (vladaionescu) [inline Earthfile:151]: This is not necessary, right?

File: Earthfile [hunk 2]
```diff
@@ -146,6 +146,16 @@ earth-docker:
...
+    COPY +earth/earth /earth-linux-amd64
+    COPY --build-arg GOOS=darwin \
+        --build-arg GOARCH=amd64 \
+        --build-arg GO_EXTRA_LDFLAGS= \
+    +earth/earth /earth-darwin-amd64
+    SAVE IMAGE --push alexcb132/earthlybinaries:latest
```
Comment (vladaionescu) [inline Earthfile:157]: I'll make an earthly repo for this.

Comment (vladaionescu) [reply Earthfile:157]: I've created this: https://hub.docker.com/repository/docker/earthly/earthlybinaries

File: Earthfile (hunk 1 above)
Comment (alexcb) [reply Earthfile:151]: good catch, I used to be using it in the tag, but no longer am and will remove it.

File: Earthfile (hunk 2 above)
Comment (alexcb) [reply Earthfile:157]: thank you kindly :)

**Comment by alexcb** [comment]
I added the `prerelease-docker` target under `all`, so the github action will automatically handle it.

**Comment by alexcb** [comment]
good point regarding the scrip, I added it to the root of the repo, let me know if that's what you meant by `./earth`

File: earth [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+NOW=`date +%s`
+since=$(expr $NOW - $LAST)
+
+if [ $since -ge 3600 ]; then
+    echo checking for latest earth pre-release binaries
+    get_latest_binary
+    echo $NOW > $LAST_CHECK_STATE_PATH
+fi
+
+~/bin/earth-bin $@
```
Comment (vladaionescu) [inline earth:29]: You probably want

```suggestion
~/bin/earth-bin "$@"
```

just in case.

File: earth [hunk 4]
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash
+set -e
+
+LAST_CHECK_STATE_PATH=/tmp/last-earth-prerelease-check
+
+function get_latest_binary {
+    docker pull alexcb132/earthlybinaries:latest
+    docker create --name earthly_binary alexcb132/earthlybinaries
```
Comment (vladaionescu) [inline earth:8]: Update to use `earthly/...` image?

File: earth [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    earthbinpath=/earth-linux-amd64
+    if [ "$(uname)" == "Darwin" ]; then
+        earthbinpath=/earth-darwin-amd64
+    fi
+
+    docker cp earthly_binary:$earthbinpath ~/bin/earth-bin
+    docker rm earthly_binary
+}
+
+LAST=`cat $LAST_CHECK_STATE_PATH 2>/dev/null || echo 0`
```
Comment (vladaionescu) [inline earth:19]:
```suggestion
LAST=`cat "$LAST_CHECK_STATE_PATH" 2>/dev/null || echo 0`
```

It's a good idea to always put variable expansions in double-quotes. Bash scripts tend to be kinda dangerous.

File: earth [hunk 6]
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash
+set -e
+
+LAST_CHECK_STATE_PATH=/tmp/last-earth-prerelease-check
```
Comment (vladaionescu) [inline earth:4]: We've generally used lower-case variable names for local vars and upper-case for exported vars in bash files. Might be good to maintain style.

File: earth [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+function get_latest_binary {
+    docker pull alexcb132/earthlybinaries:latest
+    docker create --name earthly_binary alexcb132/earthlybinaries
+
+    earthbinpath=/earth-linux-amd64
```
Comment (vladaionescu) [inline earth:10]:
```suggestion
    earth_bin_path=/earth-linux-amd64
```

Nit: style

File: earth [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    if [ "$(uname)" == "Darwin" ]; then
+        earthbinpath=/earth-darwin-amd64
+    fi
+
+    docker cp earthly_binary:$earthbinpath ~/bin/earth-bin
```
Comment (vladaionescu) [inline earth:15]:
```suggestion
    docker cp earthly_binary:"$earthbinpath" ~/bin/earth-bin
```

Just in case this script gets modified in the future.

**Comment by alexcb** [comment]
thanks for all those bash recommendations, my bash is rather sloppy.
I switched it over to use `earthly/earthlybinaries:latest` however nothing has been pushed yet, but i'm hopeful to see the github action setup perform the push.

File: earth [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+last=`cat "$last_check_state_path" 2>/dev/null || echo 0`
+now=`date +%s`
+since=$(expr $now - $last)
+
+if [ $since -ge 3600 ]; then
+    echo checking for latest earth pre-release binaries
+    get_latest_binary
+    echo $now > $last_check_state_path
```
Comment (vladaionescu) [inline earth:26]:
```suggestion
    echo "$now" > "$last_check_state_path"
```

File: earth [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    docker cp earthly_binary:"$earth_bin_path" ~/bin/earth-bin
+    docker rm earthly_binary
+}
+
+last=`cat "$last_check_state_path" 2>/dev/null || echo 0`
+now=`date +%s`
+since=$(expr $now - $last)
+
+if [ $since -ge 3600 ]; then
+    echo checking for latest earth pre-release binaries
```
Comment (vladaionescu) [inline earth:24]:
```suggestion
    echo "checking for latest earth pre-release binaries"
```

Nit: style

**Pull request merged**